### ISSUE TITLE
Registrations including associations store wrong registrations documents in database

### ISSUE THREAD

**Issue by fgalan** [telefonicaid/fiware-orion#824]
Registrations including associations store wrong registrations documents in database (empty entities and attribute arrays). Details follow.

XML case:

```
fermin@oriondeb:~$ echo 'db.dropDatabase()' | mongo orion
MongoDB shell version: 2.4.9
connecting to: orion
{ "dropped" : "orion", "ok" : 1 }
bye
fermin@oriondeb:~$ (curl localhost:1026/v1/registry/registerContext -s -S --header 'Content-Type: application/xml' -d @- | xmllint --format - ) <<EOF
<?xml version="1.0"?>
  <registerContextRequest>
    <contextRegistrationList>
      <contextRegistration>
        <registrationMetadata>
          <contextMetadata>
            <name>right_neighbour</name>
            <type>Association</type>
              <value>
                <entityAssociation>
                  <sourceEntityId type="Room" isPattern="false">
                    <id>Room1</id>
                  </sourceEntityId>
                  <targetEntityId type="Room" isPattern="false">
                    <id>Room2</id>
                  </targetEntityId>
                </entityAssociation>
                <attributeAssociationList>
                  <attributeAssociation>
                    <sourceAttribute>temperature</sourceAttribute>
                    <targetAttribute>temperature</targetAttribute>
                  </attributeAssociation>
              </attributeAssociationList>
            </value>
          </contextMetadata>
        </registrationMetadata>
        <providingApplication>http://www.fi-ware.eu/NGSI/association</providingApplication>
      </contextRegistration>
    </contextRegistrationList>
    <duration>P1M</duration>
</registerContextRequest>
EOF
<?xml version="1.0"?>
<registerContextResponse>
  <duration>P1M</duration>
  <registrationId>550060e2189109ce50b59f1f</registrationId>
</registerContextResponse>
fermin@oriondeb:~$ echo 'db.registrations.findOne()' | mongo orion
MongoDB shell version: 2.4.9
connecting to: orion
{
        "_id" : ObjectId("550060e2189109ce50b59f1f"),
        "expiration" : NumberLong(1428680162),
        "servicePath" : "/",
        "contextRegistration" : [
                {
                        "entities" : [ ],
                        "attrs" : [ ],
                        "providingApplication" : "http://www.fi-ware.eu/NGSI/association"
                }
        ]
}
bye
```

JSON case (the same) (note that in this case we use "association" instead of "attributeAssociation", given that issue https://github.com/telefonicaid/fiware-orion/issues/823 has not been fixed yet, but I guess it will happen also when it changes to "attributeAssociation").

```
fermin@oriondeb:~$ echo 'db.dropDatabase()' | mongo orion
MongoDB shell version: 2.4.9
connecting to: orion
{ "dropped" : "orion", "ok" : 1 }
bye
fermin@oriondeb:~$ (curl localhost:1026/v1/registry/registerContext -s -S --header 'Content-Type: application/json' --header 'Accept: application/json' -d @- | python -mjson.tool) <<EOF
> {
>   "contextRegistrations": [
>     {
>       "metadatas": [
>         {
>           "name": "right_neighbour",
>           "type": "Association",
>           "value": {
>             "source": {
>               "id": "Room1",
>               "type": "Room",
>               "isPattern": "false"
>             },
>             "target": {
>               "id": "Room2",
>               "type": "Room",
>               "isPattern": "false"
>             },
>             "associations": [
>               {
>                 "source": "temperature",
>                 "target": "temperature"
>               }
>             ]
>           }
>         }
>       ],
>       "providingApplication": "http://www.fi-ware.eu/NGSI/association"
>     }
>   ],
>   "duration": "P1M",
>   "registrationId": ""
> }
> EOF
{
    "duration": "P1M",
    "registrationId": "5500626c189109ce50b59f21"
}
fermin@oriondeb:~$ echo 'db.registrations.findOne()' | mongo orion
MongoDB shell version: 2.4.9
connecting to: orion
{
        "_id" : ObjectId("5500626c189109ce50b59f21"),
        "expiration" : NumberLong(1428680556),
        "servicePath" : "/",
        "contextRegistration" : [
                {
                        "entities" : [ ],
                        "attrs" : [ ],
                        "providingApplication" : "http://www.fi-ware.eu/NGSI/association"
                }
        ]
}
bye
```

**Effort: 3 man day**

**Comment by fgalan** [comment]
For reference, this is the version information of the Orion in which the above has been tested:

```
fermin@oriondeb:~$ curl --header 'Accept: application/json' localhost:1026/version
{
  "orion" : {
  "version" : "0.19.0-next",
  "uptime" : "16 d, 5 h, 41 m, 23 s",
  "git_hash" : "e62c9c66bdb56c9bebd05af54715a05aa9df4b75",
  "compile_time" : "nodate",
  "compiled_by" : "fermin",
  "compiled_in" : "oriondeb"
}
}
```

**Comment by LeandroGuillen** [comment]
After looking a bit into this, it looks more to be a philosophical matter. The fact is that this is the expected behavior, although a not very coherent with the rest of the operation set.

If we issue a registerContext request specifying both context registrations and associations it will work. For example, this is a valid request:

```
(curl localhost:1026/v1/registry/registerContext -s -S --header 'Content-Type: application/xml' -d @- | xmllint --format - ) <<EOF
<?xml version="1.0"?>
<registerContextRequest>
  <contextRegistrationList>
    <contextRegistration>
      <entityIdList>
        <entityId type="Room" isPattern="false">
          <id>ConferenceRoom</id>
        </entityId>
        <entityId type="Room" isPattern="false">
          <id>OfficeRoom</id>
        </entityId>
      </entityIdList>
      <contextRegistrationAttributeList>
        <contextRegistrationAttribute>
          <name>temperature</name>
          <type>degree</type>
          <isDomain>false</isDomain>
        </contextRegistrationAttribute>
      </contextRegistrationAttributeList>
      <registrationMetadata>
        <contextMetadata>
          <name>association1</name>
          <type>Association</type>
          <value>
            <entityAssociation>
              <sourceEntityId type="TemperatureSensor" isPattern="false">
                <id>Sensor5</id>   
              </sourceEntityId>
              <targetEntityId type="House" isPattern="false">
                <id>house_3</id>   
              </targetEntityId>
            </entityAssociation>
            <attributeAssociationList>
              <attributeAssociation>
                <sourceAttribute>measurement</sourceAttribute>
                <targetAttribute>indoorTemperature</targetAttribute>
              </attributeAssociation>
            </attributeAssociationList>
          </value>
        </contextMetadata>   
      </registrationMetadata>
      <providingApplication>http://localhost:1028/application</providingApplication>
    </contextRegistration>
  </contextRegistrationList>
  <duration>PT1H</duration>
  <registrationId></registrationId>
</registerContextRequest>
EOF
```

As a result of this:
* Two entities are registered and we can see them in the *registrations* collection on mongo
* An association is created between them and its corresponding document is created in the *associations* collection.

If we issue a request like the one this issue talks about being wrong, we will only create an association but no registration will take place. This is obvious because there is no entity information in that request outside of the association metadata.

One thing we could do to make sense of all this is to force association creation and entity registration to be independent (you cannot do them in one request).

**TL;DR** (too long; didn't read)
We have to decide whether this is how we want it to work.

**Comment by fgalan** [comment]
One alternative could be to have the `registrations` and `associations` collections independent. Thus, the following cases are possible:

* registerContext with conventional registrations only: information goes to registrations collection
* registerContext with associations medatata only: information goes to assocaitiosn collection (thus no "spurious" documents appears in the registrations collection).
* registerContext with both conventional registrations and associations metadata: each kind of information goes to the proper collection.

Anyway, this issue is not actually a bug nor urgent, so we will re-labeled as tech debt and outiside the 0.20.0 milestone.

**Comment by fgalan** [comment]
Associations (as a ConfMan feature) has been deprecated. This issue is to be closed.
